perf(ast/estree): optimize serializing strings to JSON

overlookmotel opened the pull request:
Optimize printing strings in ESTree serializer.

The main optimizations are:

1. Search for characters which need escaping in batches of 16 bytes, to speed up search over longer strings.
2. Move the initial search for characters which need escaping into a tight loop.

Additionally:

3. Tighten up the rest of the code to remove bounds checks.

overlookmotel commented:
* **#11070** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11070?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11070?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11070).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
Not much effect on `estree` benchmark, but I imagine it'd show a bigger gain on files with many longer strings (e.g. content-heavy websites).

Boshen commented:
### Merge activity

* **May 16, 11:30 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **May 16, 11:30 AM EDT**: `Boshen` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
perf(ast/estree): optimize serializing strings to JSON (#11070)

Optimize printing strings in ESTree serializer.

The main optimizations are:

1. Search for characters which need escaping in batches of 16 bytes, to speed up search over longer strings.
2. Move the initial search for characters which need escaping into a tight loop.

Additionally:

3. Tighten up the rest of the code to remove bounds checks.
